libibverbs/device.c

```diff
@@ -59,41 +59,54 @@ int __ibv_get_async_event(struct ibv_context *context,
 			  struct ibv_async_event *event);
 void __ibv_ack_async_event(struct ibv_async_event *event);
 
-static pthread_once_t device_list_once = PTHREAD_ONCE_INIT;
-static int num_devices;
-static struct ibv_device **device_list;
+static pthread_once_t init_once = PTHREAD_ONCE_INIT;
+static pthread_mutex_t dev_list_lock = PTHREAD_MUTEX_INITIALIZER;
+static int initialized;
+static struct list_head device_list = LIST_HEAD_INIT(device_list);
 
-static void count_devices(void)
+static void init_resources(void)
 {
-	num_devices = ibverbs_init(&device_list);
+	initialized = ibverbs_init();
 }
 
 struct ibv_device **__ibv_get_device_list(int *num)
 {
 	struct ibv_device **l;
-	int i;
+	struct verbs_device *device;
+	int num_devices;
+	int i = 0;
 
 	if (num)
 		*num = 0;
 
-	pthread_once(&device_list_once, count_devices);
+	pthread_once(&init_once, init_resources);
+	if (initialized < 0) {
+		errno = -initialized;
+		return NULL;
+	}
 
+	pthread_mutex_lock(&dev_list_lock);
+	num_devices = ibverbs_get_device_list(&device_list);
```

Comment:
Since we grab a mutex we don't need the pthread_once, just do the init within this mutex.